File: MergeSort.cpp
// C++ program for Merge Sort
#include <iostream>
#include<chrono>
using namespace std;
using namespace chrono;

// Merges two subarrays of array[].
// First subarray is arr[begin..mid]
// Second subarray is arr[mid+1..end]
void merge(int array[], int const left, int const mid, int const right)
{
	auto const subArrayOne = mid - left + 1;
	auto const subArrayTwo = right - mid;

	// Create temp arrays
	auto *leftArray = new int[subArrayOne],
		*rightArray = new int[subArrayTwo];

	// Copy data to temp arrays leftArray[] and rightArray[]
	for (auto i = 0; i < subArrayOne; i++)
		leftArray[i] = array[left + i];
	for (auto j = 0; j < subArrayTwo; j++)
		rightArray[j] = array[mid + 1 + j];

	auto indexOfSubArrayOne = 0, // Initial index of first sub-array
		indexOfSubArrayTwo = 0; // Initial index of second sub-array
	int indexOfMergedArray = left; // Initial index of merged array

	// Merge the temp arrays back into array[left..right]
	while (indexOfSubArrayOne < subArrayOne && indexOfSubArrayTwo < subArrayTwo) {
		if (leftArray[indexOfSubArrayOne] <= rightArray[indexOfSubArrayTwo]) {
			array[indexOfMergedArray] = leftArray[indexOfSubArrayOne];
			indexOfSubArrayOne++;
		}
		else {
			array[indexOfMergedArray] = rightArray[indexOfSubArrayTwo];
			indexOfSubArrayTwo++;
		}
		indexOfMergedArray++;
	}
	// Copy the remaining elements of
	// left[], if there are any
	while (indexOfSubArrayOne < subArrayOne) {
		array[indexOfMergedArray] = leftArray[indexOfSubArrayOne];
		indexOfSubArrayOne++;
		indexOfMergedArray++;
	}
	// Copy the remaining elements of
	// right[], if there are any
	while (indexOfSubArrayTwo < subArrayTwo) {
		array[indexOfMergedArray] = rightArray[indexOfSubArrayTwo];
		indexOfSubArrayTwo++;
		indexOfMergedArray++;
	}
}

// begin is for left index and end is
// right index of the sub-array
// of arr to be sorted */
void mergeSort(int array[], int const begin, int const end)
{
	if (begin >= end)
		return; // Returns recursively

	auto mid = begin + (end - begin) / 2;
	mergeSort(array, begin, mid);
	mergeSort(array, mid + 1, end);
	merge(array, begin, mid, end);
}

// UTILITY FUNCTIONS
// Function to print an array
void printArray(int A[], int size)
{
	for (auto i = 0; i < size; i++)
		cout << A[i] << " ";
}

// Driver code
int main()
{   
    int num;
    cout<<"Enter the number of elements in the array\n";
    cin>>num;
    int * arr = new int[num];
    srand(time(0));
	
    for(int i =0; i < num ; i++){
        arr[i] = rand();
    }

	// int arr_size = sizeof(arr) / sizeof(int);

	cout << "Given array is \n";
	
    for(int i =0 ; i < num ; i++){
        cout<<arr[i]<<" ";
    }
        cout<<endl;
        cout<<endl;
    auto start = high_resolution_clock::now();
	mergeSort(arr, 0, num- 1);
    auto stop = high_resolution_clock::now();
	cout << "\nSorted array is \n";
	printArray(arr, num);

    auto duration = duration_cast<milliseconds>(stop - start);
    cout<<"Execution time is "<<duration.count()<<" milliseconds"<<endl;
    auto nduration = duration_cast<nanoseconds>(stop-start);
    cout<<"Execution time is "<<nduration.count()<<" nanoseconds"<<endl;
    
	return 0;
}

File: MergeSort.java
import java.util.*;
public class MergeSort{
    public static void main(String[] args){
        Scanner console = new Scanner(System.in);
        System.out.println("Enter the number of elements in the array");
        int num = console.nextInt();
        Random rgn = new Random();
        
        int[] arr = new int[num];

        // System.out.println("Enter the elements in the array");        
        
        for(int i = 0; i <  arr.length; i++){
            // arr[i]=console.nextInt();
            int random = rgn.nextInt(Integer.MAX_VALUE);
            System.out.print("" + arr[i] +" ");
            arr[i] = random;
        }

        long start = System.nanoTime();
        long startms = System.currentTimeMillis();
        sort(arr,0,arr.length-1);
        long stop = System.nanoTime();
        long stopms = System.currentTimeMillis();
        long executionms = stopms - startms;
        long execution = stop -start;
        for(int i : arr){
            System.out.print("" + i +" ");
        }
        System.out.println();
        System.out.println("The total time of execution is " + execution + "ns or " + executionms + "ms");
        console.close();
    }

  public static void sort(int [] arr, int l , int r){
    if(l>=r){
        return;
    }

    int mid = l + (r-l)/2;
    sort(arr,l,mid);
    System.out.println("executed fine");
    sort(arr,mid+1,r);
    System.out.println("executed fine");
    merge(arr,l,mid,r);

  }

  private static void merge(int[] arr, int l , int m , int r){
      int leftsize=m-l+1;
      int rightsize=r-m;
      int[] leftsubarray = new int[m-l+1];
      int[] rightsubarray = new int[r-m];
      System.out.println("left size: " + leftsize + " right size: "+ rightsize);
      System.out.println("Index bound: " + m);
      for(int i = 0 ; i<leftsize; i++){
        leftsubarray[i]=arr[i+l];
      }
      System.out.println("Index bound exectured properly");
      for(int i=0; i<rightsize; i++){
          rightsubarray[i]=arr[i+m+1];
      }

      System.out.println("Index bound exectured properly");


      int leftsubarrayptr = 0;
      int rightsubarrayptr= 0;
      int mergedindex=l;
      while(leftsubarrayptr< leftsize && rightsubarrayptr<rightsize){
            arr[mergedindex++]=leftsubarray[leftsubarrayptr]<rightsubarray[rightsubarrayptr]?leftsubarray[leftsubarrayptr++]:rightsubarray[rightsubarrayptr++];
        }
        System.out.println("While loop executed both pointers");
     if(leftsubarrayptr>=leftsize){
         while(rightsubarrayptr<rightsize){
             arr[mergedindex++]=rightsubarray[rightsubarrayptr++];
         }
         System.out.println("While loop right pointer executed");
     }
    
     else{
         while(leftsubarrayptr<leftsize){
             arr[mergedindex++]=leftsubarray[leftsubarrayptr++];
         }
        System.out.println("While loop left pointer executed");

     }
}

}